### PR TITLE
TOSA: lower aten.round when followed by clamp

### PR DESCRIPTION
**Pull request by mgehre-amd**
see comment on the pattern

### CODE REVIEW

File: lib/Conversion/TorchToTosa/TorchToTosa.cpp [hunk 1]
```diff
@@ -5446,8 +5446,89 @@ LogicalResult ConvertAtenOp<AtenEmptyMemoryFormatOp>::matchAndRewrite(
...
+      //   y = clamp a, min, max
+      // so x must have a dynamic range of [min - summand, max - summand] to
+      // preserve the output value.
+      max -= summand;
+      min -= summand;
+
+      return success();
+    }
+
+    return success();
```
Comment (TinaAMD) [inline lib/Conversion/TorchToTosa/TorchToTosa.cpp:5498]: Shouldn't this be a failure? Won't we end up here when we have never set `min`/`max`?

File: lib/Conversion/TorchToTosa/TorchToTosa.cpp [hunk 2]
```diff
@@ -5446,8 +5446,89 @@ LogicalResult ConvertAtenOp<AtenEmptyMemoryFormatOp>::matchAndRewrite(
...
+// max are within the range of an i8.
+template <>
+LogicalResult ConvertAtenOp<AtenRoundOp>::matchAndRewrite(
+    AtenRoundOp op, OpAdaptor adaptor,
+    ConversionPatternRewriter &rewriter) const {
+
+  auto outputTy = getTypeConverter()
+                      ->convertType(op.getType())
+                      .dyn_cast<RankedTensorType>();
+
```
Comment (TinaAMD) [inline lib/Conversion/TorchToTosa/TorchToTosa.cpp:5514]: I think we need to constrain the type to `fp16_t`, `bf16_t` and `fp32_t`. The others do not have the expected semantics when using the tosa cast, and `AtenRound` does not seem to have any type restrictions.

File: lib/Conversion/TorchToTosa/TorchToTosa.cpp [hunk 3]
```diff
@@ -5446,8 +5446,89 @@ LogicalResult ConvertAtenOp<AtenEmptyMemoryFormatOp>::matchAndRewrite(
...
+      if (!matchPattern(add.getAlpha(), m_TorchConstantInt(&alpha)) ||
+          alpha != 1) {
+        return rewriter.notifyMatchFailure(
+            op, "Only add with alpha = 1 is handled");
+      }
+
+      int64_t summand;
+      if (!matchPattern(add.getOther(), m_TorchConstantInt(&summand))) {
+        return rewriter.notifyMatchFailure(
+            op, "Add needs constant integer second operand");
```
Comment (TinaAMD) [inline lib/Conversion/TorchToTosa/TorchToTosa.cpp:5481]:
```suggestion
            op, "Add needs constant integer as second operand");
```

File: python/torch_mlir_e2e_test/test_suite/elementwise.py [hunk 4]
```diff
@@ -3140,6 +3140,33 @@ def AtenRoundFloatModule_basic(module, tu: TestUtils):
...
+        # TOSA can only lower round when the dynamic range is restricted
+        # by a following clamp.
+        return torch.ops.aten.clamp(torch.ops.aten.round(x), -128, 127)
+
+
+@register_test_case(module_factory=lambda: AtenRoundClampFloatModule())
+def AtenRoundClampFloatModule_basic(module, tu: TestUtils):
+    module.forward(tu.rand(5, 5, low = -3.0, high = 3.0))
+
+
```
Comment (TinaAMD) [inline python/torch_mlir_e2e_test/test_suite/elementwise.py:3166]: Can you add another test which checks the float tensor + int addition case?

File: lib/Conversion/TorchToTosa/TorchToTosa.cpp [hunk 5]
```diff
@@ -5446,8 +5446,89 @@ LogicalResult ConvertAtenOp<AtenEmptyMemoryFormatOp>::matchAndRewrite(
...
+      }
+
+      if (failed(inferNecessaryRange(add, rewriter, min, max)))
+        return failure();
+
+      // The IR looks like
+      //   a = x + summand
+      //   y = clamp a, min, max
+      // so x must have a dynamic range of [min - summand, max - summand] to
```
Comment (TinaAMD) [inline lib/Conversion/TorchToTosa/TorchToTosa.cpp:5490]: Should this be `y`? Could say something like "`y` has a dynamic range of ...." as this is not describing a condition, but rather explaining a fact.

**Review by TinaAMD — CHANGES_REQUESTED** [review]

File: lib/Conversion/TorchToTosa/TorchToTosa.cpp [hunk 6]
```diff
@@ -5446,8 +5446,89 @@ LogicalResult ConvertAtenOp<AtenEmptyMemoryFormatOp>::matchAndRewrite(
...
+
+  int64_t min, max;
+  if (failed(inferNecessaryRange(op, rewriter, &min, &max)))
+    return failure();
+
+  Type castToType;
+  if (min >= -128 && max <= 127) {
+    castToType = rewriter.getIntegerType(8);
+  } else {
+    return rewriter.notifyMatchFailure(op, "currently only supporting i8");
```
[the remark this answers is not in the document]
Comment (mgehre-amd) [reply lib/Conversion/TorchToTosa/TorchToTosa.cpp:5523]: I wanted to write it this way to have the right structure for adding more int types later on.

File: lib/Conversion/TorchToTosa/TorchToTosa.cpp (hunk 1 above)
Comment (mgehre-amd) [reply lib/Conversion/TorchToTosa/TorchToTosa.cpp:5498]: correct!

File: lib/Conversion/TorchToTosa/TorchToTosa.cpp (hunk 5 above)
Comment (mgehre-amd) [reply lib/Conversion/TorchToTosa/TorchToTosa.cpp:5490]: I updated this to say
```
      // The IR looks like
      //   a = x + summand
      //   y = clamp a, min, max
      // which is the same as
      //   c = clamp a, min - summand, max - summand
      //   y = c + summand.
```

File: python/torch_mlir_e2e_test/test_suite/elementwise.py (hunk 4 above)
Comment (mgehre-amd) [reply python/torch_mlir_e2e_test/test_suite/elementwise.py:3166]: will do

File: lib/Conversion/TorchToTosa/TorchToTosa.cpp (hunk 3 above)
Comment (mgehre-amd) [reply lib/Conversion/TorchToTosa/TorchToTosa.cpp:5481]: done

File: lib/Conversion/TorchToTosa/TorchToTosa.cpp (hunk 2 above)
Comment (mgehre-amd) [reply lib/Conversion/TorchToTosa/TorchToTosa.cpp:5514]: Good catch! Would be better if torch-mlir would allow lit testing for all those cases...

**Review by TinaAMD — APPROVED** [review]

**Pull request closed without merging**